test/e2e/fqdn_dns_cache_test.go

```diff
@@ -0,0 +1,352 @@
+	require.NoError(t, data.podWaitForRunning(defaultTimeout, customDnsPodName, data.testNamespace))
+}
+
+func createFqdnPolicyInNamespace(t *testing.T, data *TestData, domainName string) {
+	const fqdnPolicyName = "test-anp-fqdn"
+	podSelectorLabel := map[string]string{
+		fqdnPodSelectorLabelKey: fqdnPodSelectorLabelValue,
+	}
+	port := int32(80)
+	udpPort := int32(53)
+	builder := &utils.AntreaNetworkPolicySpecBuilder{}
+	builder = builder.SetName(data.testNamespace, fqdnPolicyName).
+		SetTier(defaultTierName).
+		SetPriority(1.0).
+		SetAppliedToGroup([]utils.ANNPAppliedToSpec{{PodSelector: podSelectorLabel}})
+	builder.AddFQDNRule(domainName, utils.ProtocolTCP, &port, nil, nil, "AllowForFQDN", nil,
+		crdv1beta1.RuleActionAllow)
+	builder.AddEgress(utils.ProtocolUDP, &udpPort, nil, nil, nil, nil,
+		nil, nil, nil, nil, map[string]string{customDnsLabelKey: customDnsLabelValue},
+		nil, nil, nil, nil,
+		nil, nil, crdv1beta1.RuleActionAllow, "", "AllowDnsQueries")
+	builder.AddEgress(utils.ProtocolTCP, nil, nil, nil, nil, nil,
+		nil, nil, nil, nil, nil,
+		nil, nil, nil, nil,
+		nil, nil, crdv1beta1.RuleActionReject, "", "DropAllRemainingTraffic")
+
+	annp, err := k8sUtils.CreateOrUpdateANNP(builder.Get())
+	require.NoError(t, err, "error while deploying antrea policy %+v", err)
+	require.NoError(t, waitForResourceReady(t, 30*time.Second, annp))
+}
+
+func createToolBoxPod(t *testing.T, data *TestData, dnsServiceIP string) {
+	toolBoxLabel := map[string]string{fqdnPodSelectorLabelKey: fqdnPodSelectorLabelValue}
+	mutateSpecForAddingCustomDNS := func(pod *corev1.Pod) {
+		pod.Spec.DNSPolicy = corev1.DNSNone
+		if pod.Spec.DNSConfig == nil {
+			pod.Spec.DNSConfig = &corev1.PodDNSConfig{}
+		}
+		pod.Spec.DNSConfig.Nameservers = []string{dnsServiceIP}
+
+	}
+	require.NoError(t, NewPodBuilder(toolboxPodName, data.testNamespace, ToolboxImage).WithLabels(toolBoxLabel).WithContainerName(toolboxContainerName).WithMutateFunc(mutateSpecForAddingCustomDNS).Create(data))
+	require.NoError(t, data.podWaitForRunning(defaultTimeout, toolboxPodName, data.testNamespace))
+}
+
+func createHttpAgnhostPod(t *testing.T, data *TestData, podName string, agnLabels map[string]string) *PodIPs {
+	const agnHostPort = 80
+	args := []string{"netexec", "--http-port=" + strconv.Itoa(agnHostPort)}
+	ports := []corev1.ContainerPort{
+		{
+			Name:          "http",
+			ContainerPort: agnHostPort,
+			Protocol:      corev1.ProtocolTCP,
+		},
+	}
+
+	require.NoError(t, NewPodBuilder(podName, data.testNamespace, agnhostImage).WithArgs(args).WithPorts(ports).WithLabels(agnLabels).Create(data))
+	podIPs, err := data.podWaitForIPs(defaultTimeout, podName, data.testNamespace)
+	require.NoError(t, err)
+	return podIPs
+}
+
+func extractIPs(podIPs *PodIPs) (string, string) {
+	var ipv4, ipv6 string
+	res := podIPs.AsSlice()
+	if len(res) > 0 {
+		ipv4 = res[0].String()
+		if len(res) == 2 {
+			ipv6 = res[1].String()
+		}
+	}
+	return ipv4, ipv6
+}
```

Comment:
there is no need to call `AsSlice()` in this function

let's replace this with a new method for `PodIPs`: `func (p *PodIPs) AsStrings() (string, string)`

it may be helpful for other tests, and then we don't need to define this top-level function here